### PR TITLE
Apply new logging system from #2677 project-wise

### PR DESCRIPTION
**Pull request by junpenglao**
1, add release note for #2677
2, change `pm._log` --> `_log` from logging
3, change `warnings.warn` --> `_log.warning`, remove `import warnings`

### CODE REVIEW

**Comment by aloctavodia** [comment]
This will introduce a bug in the `compare` function. These [lines](https://github.com/junpenglao/pymc3/blob/e3e074d7b816c0cfa4b82823937596c19c9e7733/pymc3/stats.py#L578-L588) capture the warnings generated by `waic` or `loo`, (which are then included in the resulting DataFrame).

**Comment by aseyboldt** [comment]
I think some of those should still be normal warnings. The [docs](https://docs.python.org/2/howto/logging.html#when-to-use-logging) on that topic:

> warnings.warn() in library code if the issue is avoidable and the client application should be modified to eliminate the warning

> logging.warning() if there is nothing the client application can do about the situation, but the event should still be noted

I'll mark the one where I think the first would be appropriate in the code...

File: pymc3/backends/base.py [hunk 1]
```diff
@@ -310,7 +308,7 @@ def __getitem__(self, idx):
         var = str(var)
         if var in self.varnames:
             if var in self.stat_names:
-                warnings.warn("Attribute access on a trace object is ambigous. "
+                _log.warning("Attribute access on a trace object is ambigous. "
```
Comment (aseyboldt) [inline pymc3/backends/base.py:311]: `warnings.warn`

File: pymc3/backends/base.py [hunk 2]
```diff
@@ -330,7 +328,7 @@ def __getattr__(self, name):
         name = str(name)
         if name in self.varnames:
             if name in self.stat_names:
-                warnings.warn("Attribute access on a trace object is ambigous. "
+                _log.warning("Attribute access on a trace object is ambigous. "
```
Comment (aseyboldt) [inline pymc3/backends/base.py:331]: `warnings.warn`

File: pymc3/backends/base.py [hunk 3]
```diff
@@ -384,7 +382,7 @@ def add_values(self, vals):
             l_samples = len(self) * len(self.chains)
             l_v = len(v)
             if l_v != l_samples:
-                warnings.warn("The lenght of the values you are trying to "
+                _log.warning("The lenght of the values you are trying to "
```
Comment (aseyboldt) [inline pymc3/backends/base.py]: `warnings.warn`

File: pymc3/backends/smc_text.py [hunk 4]
```diff
@@ -240,7 +243,7 @@ def load_multitrace(self, stage_number, model=None):
         dirname = self.stage_path(stage_number)
         files = glob(os.path.join(dirname, 'chain_*.csv'))
         if len(files) < 1:
-            pm._log.warn('Stage directory %s contains no traces!' % dirname)
+            _log.warning('Stage directory %s contains no traces!' % dirname)
```
Comment (aseyboldt) [inline pymc3/backends/smc_text.py:246]: Shouldn't this actually be an exception?

File: pymc3/backends/smc_text.py [hunk 5]
```diff
@@ -272,7 +275,7 @@ def check_multitrace(self, mtrace, draws, n_chains):
         for chain in range(n_chains):
             if chain in mtrace.chains:
                 if len(mtrace._straces[chain]) != draws:
-                    pm._log.warn('Trace number %i incomplete' % chain)
+                    _log.warning('Trace number %i incomplete' % chain)
```
Comment (aseyboldt) [inline pymc3/backends/smc_text.py:278]: exception?

File: pymc3/backends/smc_text.py [hunk 6]
```diff
@@ -412,11 +415,11 @@ def _load_df(self):
             try:
                 self.df = pd.read_csv(self.filename)
             except pd.parser.EmptyDataError:
-                pm._log.warn('Trace %s is empty and needs to be resampled!' % self.filename)
+                _log.warning('Trace %s is empty and needs to be resampled!' % self.filename)
```
Comment (aseyboldt) [inline pymc3/backends/smc_text.py:418]: exception?

File: pymc3/backends/smc_text.py [hunk 7]
```diff
@@ -412,11 +415,11 @@ def _load_df(self):
...
                 os.remove(self.filename)
                 self.corrupted_flag = True
             except pd.io.common.CParserError:
-                pm._log.warn('Trace %s has wrong size!' % self.filename)
+                _log.warning('Trace %s has wrong size!' % self.filename)
```
Comment (aseyboldt) [inline pymc3/backends/smc_text.py:422]: exception?
Are there use cases where we'd want the code to continue here?

File: pymc3/distributions/continuous.py [hunk 8]
```diff
@@ -64,7 +65,7 @@ def assert_negative_support(var, label, distname, value=-1e-6):
     if np.any(support):
         msg = "The variable specified for {0} has negative support for {1}, ".format(label, distname)
         msg += "likely making it unsuitable for this parameter."
-        warnings.warn(msg)
+        _log.warning(msg)
```
Comment (aseyboldt) [inline pymc3/distributions/continuous.py:67]: `warnings.warn`
because the code should be fixed.

File: pymc3/distributions/discrete.py [hunk 9]
```diff
@@ -564,7 +565,7 @@ class Constant(Discrete):
     """
 
     def __init__(self, c, *args, **kwargs):
-        warnings.warn("Constant has been deprecated. We recommend using a Determinstic object instead.",
+        _log.warning("Constant has been deprecated. We recommend using a Determinstic object instead.",
```
Comment (aseyboldt) [inline pymc3/distributions/discrete.py:762]: `warnings.warn`

File: pymc3/distributions/multivariate.py [hunk 10]
```diff
@@ -660,7 +662,7 @@ class Wishart(Continuous):
 
     def __init__(self, nu, V, *args, **kwargs):
         super(Wishart, self).__init__(*args, **kwargs)
-        warnings.warn('The Wishart distribution can currently not be used '
+        _log.warning('The Wishart distribution can currently not be used '
```
Comment (aseyboldt) [inline pymc3/distributions/multivariate.py:668]: `warnings.warn`

File: pymc3/distributions/multivariate.py [hunk 11]
```diff
@@ -1016,7 +1018,7 @@ class LKJCorr(Continuous):
 
     def __init__(self, eta=None, n=None, p=None, transform='interval', *args, **kwargs):
         if (p is not None) and (n is not None) and (eta is None):
-            warnings.warn('Parameters to LKJCorr have changed: shape parameter n -> eta '
+            _log.warning('Parameters to LKJCorr have changed: shape parameter n -> eta '
```
Comment (aseyboldt) [inline pymc3/distributions/multivariate.py:1024]: `warnings.warn`

File: pymc3/distributions/multivariate.py [hunk 12]
```diff
@@ -1039,7 +1041,7 @@ def __init__(self, eta=None, n=None, p=None, transform='interval', *args, **kwar
 
         super(LKJCorr, self).__init__(shape=shape, transform=transform,
                                       *args, **kwargs)
-        warnings.warn('Parameters in LKJCorr have been rename: shape parameter n -> eta '
+        _log.warning('Parameters in LKJCorr have been rename: shape parameter n -> eta '
```
Comment (aseyboldt) [inline pymc3/distributions/multivariate.py:1047]: `warnings.warn`

File: pymc3/plots/energyplot.py [hunk 13]
```diff
@@ -44,7 +45,7 @@ def energyplot(trace, kind='kde', figsize=None, ax=None, legend=True,
     try:
         energy = trace['energy']
     except KeyError:
-        warnings.warn('There is no energy information in the passed trace.')
+        _log.warning('There is no energy information in the passed trace.')
```
Comment (aseyboldt) [inline pymc3/plots/energyplot.py:52]: Shouldn't this be an exception as well?

File: pymc3/sampling.py [hunk 14]
```diff
@@ -344,7 +343,7 @@ def sample(draws=500, step=None, init='auto', n_init=200000, start=None,
 
     if 'chain' in kwargs:
         chain_idx = kwargs['chain']
-        warnings.warn(
+        _log.warning(
```
Comment (aseyboldt) [inline pymc3/sampling.py:360]: `warnings.warn`

File: pymc3/stats.py [hunk 15]
```diff
@@ -115,8 +120,7 @@ def dic(trace, model=None):
     z : float
         The deviance information criterion of the model and trace
     """
-    warnings.warn("dic has been deprecated. Use `waic` or `loo` instead.", DeprecationWarning,
-                  stacklevel=2)
+    _log.warning("dic has been deprecated. Use `waic` or `loo` instead.")
```
Comment (aseyboldt) [inline pymc3/stats.py]: `warnings.warn`

File: pymc3/stats.py [hunk 16]
```diff
@@ -216,7 +220,7 @@ def waic(trace, model=None, pointwise=False, progressbar=False):
 
     vars_lpd = np.var(log_py, axis=0)
     if np.any(vars_lpd > 0.4):
-        warnings.warn("""For one or more samples the posterior variance of the
+        _log.warning("""For one or more samples the posterior variance of the
```
Comment (aseyboldt) [inline pymc3/stats.py:212]: Really not sure here. We might also want to change this to `backend.report.SamplerWarning`. The functionally should match, only the name seems a bit off...
Which maybe indicates that the name `SamplerWarning` isn't right.

File: pymc3/stats.py [hunk 17]
```diff
@@ -284,7 +288,7 @@ def loo(trace, model=None, pointwise=False, reff=None, progressbar=False):
     lw, ks = _psislw(-log_py, reff)
     lw += log_py
     if np.any(ks > 0.7):
-        warnings.warn("""Estimated shape parameter of Pareto distribution is
+        _log.warning("""Estimated shape parameter of Pareto distribution is
```
Comment (aseyboldt) [inline pymc3/stats.py:291]: same

File: pymc3/stats.py [hunk 18]
```diff
@@ -466,8 +470,7 @@ def bpic(trace, model=None):
     z : float
         The Bayesian predictive information criterion of the model and trace
     """
-    warnings.warn("bpic has been deprecated. Use `waic` or `loo` instead.", DeprecationWarning,
-                  stacklevel=2)
+    _log.warning("bpic has been deprecated. Use `waic` or `loo` instead.")
```
Comment (aseyboldt) [inline pymc3/stats.py]: `warnings.warn`

File: pymc3/stats.py [hunk 19]
```diff
@@ -858,7 +861,7 @@ def quantiles(x, qlist=(2.5, 25, 50, 75, 97.5), transform=lambda x: x):
         return dict(zip(qlist, quants))
 
     except IndexError:
-        pm._log.warning("Too few elements for quantile calculation")
+        _log.warning("Too few elements for quantile calculation")
```
Comment (aseyboldt) [inline pymc3/stats.py:833]: Why do we catch this exception? We just return `None` here. Maybe this should be
`raise ValueError('Too few elements for quantile calculation')`?

File: pymc3/stats.py [hunk 20]
```diff
@@ -1011,8 +1014,7 @@ def summary(trace, varnames=None, transform=lambda x: x, stat_funcs=None,
 
 
 def df_summary(*args, **kwargs):
-    warnings.warn("df_summary has been deprecated. In future, use summary instead.",
-                  DeprecationWarning, stacklevel=2)
+    _log.warning("df_summary has been deprecated. In future, use summary instead.")
```
Comment (aseyboldt) [inline pymc3/stats.py]: `warnings.warn`

File: pymc3/tuning/starting.py [hunk 21]
```diff
@@ -80,7 +84,7 @@ def find_MAP(start=None, vars=None, method="L-BFGS-B",
         compute_gradient = False
 
     if disc_vars or not compute_gradient:
-        pm._log.warning("Warning: gradient not available." +
+        _log.warning("Warning: gradient not available." +
```
Comment (aseyboldt) [inline pymc3/tuning/starting.py:87]: not sure about this one...

File: pymc3/tuning/starting.py [hunk 22]
```diff
@@ -89,7 +93,7 @@ def find_MAP(start=None, vars=None, method="L-BFGS-B",
 
     if "fmin" in kwargs:
         fmin = kwargs.pop("fmin")
-        warnings.warn('In future versions, set the optimization algorithm with a string. '
+        _log.warning('In future versions, set the optimization algorithm with a string. '
```
Comment (aseyboldt) [inline pymc3/tuning/starting.py:96]: I guess this is `warnings.warn`, DeprecationWarning?

File: pymc3/variational/inference.py [hunk 23]
```diff
@@ -58,7 +57,7 @@ def _maybe_score(self, score):
         if score is None:
             score = returns_loss
         elif score and not returns_loss:
-            warnings.warn('method `fit` got `score == True` but %s '
+            _log.warning('method `fit` got `score == True` but %s '
```
Comment (aseyboldt) [inline pymc3/variational/inference.py:60]: `warnings.warn` or exception.
@ferrine Why isn't this an exception?

File: pymc3/variational/opvi.py [hunk 24]
```diff
@@ -141,7 +143,7 @@ class ObjectiveUpdates(theano.OrderedUpdates):
 
 
 def _warn_not_used(smth, where):
-    warnings.warn('`%s` is not used for %s and ignored' % (smth, where))
+    _log.warning('`%s` is not used for %s and ignored' % (smth, where))
```
Comment (aseyboldt) [inline pymc3/variational/opvi.py:146]: Again, why not an exception?

File: pymc3/variational/opvi.py [hunk 25]
```diff
@@ -423,7 +425,7 @@ def __call__(self, f=None):
                     f = TestFunction.from_function(f)
         else:
             if f is not None:
-                warnings.warn(
+                _log.warning(
```
Comment (aseyboldt) [inline pymc3/variational/opvi.py:428]: `warnings.warn`

**Comment by aseyboldt** [comment]
Sorry for the long list :-)
Or do you think we should stick with just `logging.warn`?

**Comment by junpenglao** [comment]
Thanks for the careful reading ;-)
I think maybe it is better to use `logging.warn`, at least from the user point of view, they are seeing `warning` either case, and it is still the user's judgement and decision whether they are going to take action or not.

File: pymc3/backends/base.py (hunk 3 above)
Comment (ColCarroll) [reply pymc3/backends/base.py]: also a typo! `lenght` --> `length`

**Comment by twiecki** [comment]
Closing because of age. Feel free to reopen if you want to fix this up.

**Pull request closed without merging**